cpp/src/arrow/adapters/orc/jni/jni_wrapper.cpp

```diff
@@ -0,0 +1,243 @@
+    ThrowJavaException(env, "Unable to get java class field: " + sig);
+  }
+
+  return ret;
+}
+
+std::unique_ptr<ORCFileReader>* GetNativeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj, GetFieldId(env, env->GetObjectClass(this_obj), "nativeReaderAddress"));
+  return reinterpret_cast<std::unique_ptr<ORCFileReader>*>(reader);
+}
+
+std::shared_ptr<RecordBatchReader>* GetStripeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj,
+      GetFieldId(env, env->GetObjectClass(this_obj), "nativeStripeReaderAddress"));
+  return reinterpret_cast<std::shared_ptr<RecordBatchReader>*>(reader);
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_open(
+    JNIEnv* env, jobject this_obj, jstring file_path) {
+  std::shared_ptr<arrow::io::ReadableFile> in_file;
+  const char* str = env->GetStringUTFChars(file_path, nullptr);
+  std::string path(str);
+  env->ReleaseStringUTFChars(file_path, str);
+
+  arrow::Status ret;
+  if (path.find("hdfs://") == 0) {
+    return false;
+  } else {
+    ret = arrow::io::ReadableFile::Open(path, &in_file);
+  }
+
+  if (ret.ok()) {
+    auto reader = new std::unique_ptr<ORCFileReader>();
+
+    ret = ORCFileReader::Open(
+        std::static_pointer_cast<arrow::io::RandomAccessFile>(in_file),
+        arrow::default_memory_pool(), reader);
+
+    if (!ret.ok()) {
+      delete reader;
+      ThrowJavaIOException(env, std::invalid_argument("Failed open file" + path));
+    }
+
+    env->SetLongField(
+        this_obj, GetFieldId(env, env->GetObjectClass(this_obj), "nativeReaderAddress"),
+        reinterpret_cast<long>(reader));
+  }
+
+  return ret.ok();
+}
+
+JNIEXPORT void JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_close(
+    JNIEnv* env, jobject this_obj) {
+  delete GetNativeReader(env, this_obj);
+}
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_seek(
+    JNIEnv* env, jobject this_obj, jint row_number) {
+  auto reader = GetNativeReader(env, this_obj);
+  return (*reader)->Seek(row_number).ok();
+}
+
+JNIEXPORT jint JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_getNumberOfStripes(
+    JNIEnv* env, jobject this_obj) {
+  auto reader = GetNativeReader(env, this_obj);
+  return (*reader)->NumberOfStripes();
+}
+
+JNIEXPORT jobject JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_nextStripeReader(JNIEnv* env,
+                                                                       jobject this_obj,
+                                                                       jlong batch_size) {
+  auto reader = GetNativeReader(env, this_obj);
+  auto stripe_reader = new std::shared_ptr<RecordBatchReader>();
+  auto status = (*reader)->NextStripeReader(batch_size, stripe_reader);
+
+  jobject ret = nullptr;
+  if (!status.ok()) {
+    delete stripe_reader;
```

Comment:
if you wrap the the allocation in a unique_ptr, and then call release on it below, it might be safer then manually deleting